File: src/ruby/tormenta20/models/encantamento.rb
# frozen_string_literal: true

require "active_record"
require "json"

module Tormenta20
  module Models
    # Model representing an Encantamento (Enchantment) in Tormenta20.
    #
    # Enchantments are magical properties applied to permanent magic items
    # (weapons, armor, shields). They are distinct from Melhorias (enhancements),
    # which apply to superior items.
    #
    # @example Find all weapon enchantments
    #   Tormenta20.encantamentos.where(categoria: "arma")
    #
    # @example Find all armor/shield enchantments
    #   Tormenta20.encantamentos.where(categoria: "armadura_escudo")
    class Encantamento < Base
      self.table_name = "encantamentos"

      validates :id, presence: true, uniqueness: true
      validates :name, presence: true
      validates :categoria, presence: true, inclusion: { in: %w[arma armadura_escudo] }

      scope :armas, -> { where(categoria: "arma") }
      scope :armaduras_escudos, -> { where(categoria: "armadura_escudo") }
      scope :escudo_only, -> { where(escudo_only: true) }

      def to_h
        {
          id: id,
          name: name,
          description: description,
          categoria: categoria,
          escudo_only: escudo_only == 1,
          conta_como_dois: conta_como_dois == 1,
          prerequisitos: prerequisitos,
          efeito: efeito
        }.compact
      end
    end
  end
end

File: src/ts/models/encantamento.ts
import { BaseModel } from "./base.js"
import { Query } from "../query.js"

const TABLE = "encantamentos"

/** Whether the enchantment applies to weapons or to armors/shields. */
export type EncantamentoCategoria = "arma" | "armadura"

/**
 * An Encantamento (magical enchantment) that can be applied to weapons or armor/shields.
 *
 * @example
 * ```ts
 * Encantamento.armas().all()
 * Encantamento.armadurasEscudos().all()
 * Encantamento.escudoOnly().all()
 * Encantamento.find("afiado")
 * ```
 */
export class Encantamento extends BaseModel {
  get description(): string | null { return (this._row.description as string) ?? null }
  get categoria(): EncantamentoCategoria { return this._row.categoria as EncantamentoCategoria }
  /** `true` if this enchantment can only be applied to shields (not regular armors). */
  get escudoOnly(): boolean { return BaseModel.bool(this._row.escudo_only) }
  /** `true` if this enchantment counts as two enchantment slots. */
  get contaComoDois(): boolean { return BaseModel.bool(this._row.conta_como_dois) }
  /** JSON array of prerequisite descriptors. */
  get prerequisitos(): unknown[] { return BaseModel.json(this._row.prerequisitos, []) }
  /** JSON object with mechanical effects. */
  get efeito(): Record<string, unknown> { return BaseModel.json(this._row.efeito, {}) }

  toH() {
    return {
      id: this.id, name: this.name, description: this.description, categoria: this.categoria,
      escudo_only: this.escudoOnly, conta_como_dois: this.contaComoDois,
      prerequisitos: this.prerequisitos, efeito: this.efeito,
    }
  }

  static query(): Query<Encantamento> { return BaseModel.makeQuery(TABLE, Encantamento) }
  static all(): Encantamento[] { return Encantamento.query().all() }
  static find(id: string): Encantamento | null { return Encantamento.query().find(id) }
  static first(): Encantamento | null { return Encantamento.query().first() }
  static last(): Encantamento | null { return Encantamento.query().last() }
  static count(): number { return Encantamento.query().count() }

  /** Returns enchantments that apply to weapons. */
  static armas(): Query<Encantamento> { return Encantamento.query().where("categoria = ?", "arma") }
  /** Returns enchantments that apply to armors and shields. */
  static armadurasEscudos(): Query<Encantamento> { return Encantamento.query().where("categoria = ?", "armadura") }
  /** Returns armor/shield enchantments applicable only to shields. */
  static escudoOnly(): Query<Encantamento> { return Encantamento.query().where("escudo_only = 1") }
}
